cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java

```diff
@@ -95,14 +96,35 @@ protected final HttpServiceRequest wrapRequest(HttpRequest request) {
    * object. This method is called from handler class generated by {@link HttpHandlerGenerator}.
    */
   @SuppressWarnings("unused")
-  protected final DelayedHttpServiceResponder wrapResponder(HttpResponder responder) {
+  protected final DelayedHttpServiceResponder wrapResponder(HttpResponder responder,
+                                                            final TransactionContext txContext) {
+
+
     MetricsContext collector = this.metricsContext;
     HttpServiceContext serviceContext = context.getServiceContext();
-    if (serviceContext != null && serviceContext.getSpecification() != null) {
+    Preconditions.checkState(serviceContext instanceof TransactionalHttpServiceContext,
+                             "This instance of HttpServiceContext does not support transactions.");
+
+    if (serviceContext.getSpecification() != null) {
       collector = metricsContext.childContext(Constants.Metrics.Tag.HANDLER,
                                               serviceContext.getSpecification().getName());
     }
-    return new DelayedHttpServiceResponder(responder, collector);
+
+    return new DelayedHttpServiceResponder(responder, new BodyProducerFactory() {
+      @Override
+      public BodyProducer create(HttpContentProducer contentProducer, TransactionalHttpServiceContext serviceContext) {
+        final ClassLoader programContextClassLoader = new CombineClassLoader(
+          null, ImmutableList.of(contentProducer.getClass().getClassLoader(), getClass().getClassLoader()));
+
+        Transactional transactional = createTransactional(txContext, programContextClassLoader, serviceContext);
+
+        // Capture the context since we need to keep it till the end of the content producing.
+        // This method only gets call when HttpContentConsumer is not used
```

Comment:
When `HttpContentConsumer` is used, a different `BodyProducerFactory` will be used, hence this code path will not be executed.